Solver - Reviewer Round Robin

justinpayan opened the pull request:
This branch implements the Greedy Reviewer Round Robin algorithm of https://arxiv.org/pdf/2108.02126.pdf, modified to use the picking sequence from https://ifaamas.org/Proceedings/aamas2020/pdfs/p231.pdf. 

The returned reviewer assignment is weighted envy free up to 1 item (WEF1). An envy-free (EF) allocation would ensure no paper prefers the reviewers assigned to another paper - weighted EF1 is a relaxation of envy-freeness that applies when papers have uneven reviewer demands. The assignment algorithm can also support min/max reviewer supply constraints.

melisabok commented:
Is this failing because we need to have this PR merge first? https://github.com/openreview/openreview-py/pull/1168

melisabok commented on setup.py:
```diff
@@ -18,6 +18,7 @@
         "openreview-py",
         "ortools>=8.1.8487",
         "pytest",
+        "sortedcontainers>=2.4.0",
```
what is this new library for?

purujitgoyal replied:
The new algorithm.

justinpayan replied:
The algorithm needs to maintain a sorted list of papers in order of priority, because at each step there will usually be multiple papers with top priority and we need to iterate over them. I used the SortedList class from this library to implement that.

purujitgoyal commented:
> Is this failing because we need to have this PR merge first? [openreview/openreview-py#1168](https://github.com/openreview/openreview-py/pull/1168)

Yes, I am planning to move tests from circleCI to github for the future.

melisabok commented on tests/test_solvers_fairflow.py:
```diff
@@ -493,12 +493,12 @@ def test_solver_fairflow_respect_minimums_3():
+            if res[pix, rix] != 0:
                 reviewer_count_reviews += 1
-        assert reviewer_count_reviews >= 1
+        assert reviewer_count_reviews >= 2
```
why did you have to change a FairFlow test in this PR?

purujitgoyal replied:
@justinpayan

justinpayan replied:
I changed the assert because above, on line 560, min_papers_per_reviewer is set to 2. So it seemed like the test was actually meant to be checking if each reviewer got at least 2 papers.

justinpayan replied:
I will absolutely not mind if you don't want to merge these test suite changes, I just thought I'd suggest them.

melisabok commented on tests/test_integration_fairflow.py:
```diff
@@ -115,7 +115,7 @@ def test_integration_supply_mismatch_error(
     test_client = openreview_context["test_client"]
 
-    conference_id = "AKBC.ws/2019/Conference"
+    conference_id = "AKBC.ws/2049/Conference"
```
same here, why did you have to change this test?

purujitgoyal replied:
This is some old change I made. I believe there was already a conference with id `AKBC.ws/2019/Conference`.

justinpayan commented on tests/test_solvers_fairflow.py:
```diff
@@ -185,7 +185,7 @@ def test_solver_fairflow_no_0_score_assignment():
     Reviewers review min: 1, max: 3 papers.
     Each paper needs 2 reviews.
-    No constraints.
+    Reviewer 0 cannot review paper 1.
```
The constraint matrix inside the test has a -1 in element [0, 1], so there actually is this constraint.

justinpayan commented on tests/test_solvers_fairflow.py:
```diff
@@ -449,10 +449,10 @@ def test_solver_fairflow_respect_minimums_2():
-        for pix in range(ncols):
-            if res[rix, pix] != 0:
+        for pix in range(nrows):
+            if res[pix, rix] != 0:
```
All 3 of these identical changes are because semantically, the returned matrix from the solver has papers in rows and reviewers in columns. So pix (I assume standing for paper index) should index the rows and rix (reviewer index) should index the columns. It doesn't change any logic but I think it's more semantically accurate.

justinpayan commented on tests/test_solvers_fairflow.py:
```diff
@@ -582,12 +582,12 @@ def test_solver_fairflow_respects_two_minimum():
+            if res[pix, rix] != 0:
                 reviewer_count_reviews += 1
-        assert reviewer_count_reviews >= 1
+        assert reviewer_count_reviews >= 2
```
Line 560 sets min_papers_per_reviewer to 2.

melisabok commented:
I'm testing this with real date and I noticed the `min_papers` constraint is not being taken into account:
<img width="892" alt="Screen Shot 2022-04-08 at 2 47 43 PM" src="https://user-images.githubusercontent.com/545506/162503168-dab461cb-5929-4a65-9152-c03d6aab07de.png">
<img width="603" alt="Screen Shot 2022-04-08 at 2 48 02 PM" src="https://user-images.githubusercontent.com/545506/162503179-daca5c99-501d-4fdf-838b-bc0358af4728.png">

When I try with `min_papers` = 6 or 7 I get a `No Solution` error what is expected but with `min_papers` = 4 or 5 I get a solution but there are users with 3, 2, 1 and 0 assigned papers and that shouldn't happen.

Do you have a test that checks the `min_papers` constraint is respected?

justinpayan commented:
Hi Melisa,

Thanks for running the test. That's very strange.

The algorithm should check, after each individual paper-reviewer
assignment, if the total remaining demand for papers equals the total
difference between reviewers' assignments and lower bounds. If so, it will
only assign papers to reviewers that still haven't met their lower bound.
This happens here:
https://github.com/openreview/openreview-matcher/blob/ea34bb882aa20df1cde4805f53605f1b52f0dca1/matcher/solvers/fairsequence.py#L343
.

I have 5 tests for this condition (starting at
https://github.com/openreview/openreview-matcher/blob/ea34bb882aa20df1cde4805f53605f1b52f0dca1/tests/test_solvers_fairsequence.py#L569
).

Is there any way I can debug the algorithm on this data? If not, maybe you
could tell me the number of reviewers and papers and I could try to
reproduce the behavior on a synthetic dataset?

Thanks,
Justin

On Fri, Apr 8, 2022 at 2:51 PM Melisa Bok ***@***.***> wrote:

> I'm testing this with real date and I noticed the min_papers constraint
> is not being taken into account:
> [image: Screen Shot 2022-04-08 at 2 47 43 PM]
> <https://user-images.githubusercontent.com/545506/162503168-dab461cb-5929-4a65-9152-c03d6aab07de.png>
> [image: Screen Shot 2022-04-08 at 2 48 02 PM]
> <https://user-images.githubusercontent.com/545506/162503179-daca5c99-501d-4fdf-838b-bc0358af4728.png>
>
> When I try with min_papers = 6 or 7 I get a No Solution error what is
> expected but with min_papers = 4 or 5 I get a solution but there are
> users with 3, 2, 1 and 0 assigned papers and that shouldn't happen.
>
> Do you have a test that checks the min_papers constraint is respected?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openreview/openreview-matcher/pull/226#issuecomment-1093230540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDA66NZ2HYKSWYWNZFL2QTVEB52TANCNFSM5ER5C7XQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Justin Payan

PhD Candidate

College of Information and Computer Sciences

University of Massachusetts Amherst

melisabok commented:
Hi @justinpayan ,

Apologies for this. This is my mistake and I didn't notice there were users with a reduce quota lower than the global `min_papers` value.

However with @purujitgoyal we were testing ICLR AC assignment and we couldn't find a solution using FairSequence while we found a solution using FairFlow and MinMax solvers. We don't know how to explain that. 

Something that we found in the logs is that the following message:

```
Unable to find a WEF1 allocation satisfying all papers' demands. Falling back to picking sequence without WEF1 guarantees.
```

Purujit can reproduce in his local environment so you may want to have a call and debug this together.

Thanks,

Melisa

justinpayan commented:
Hi Melisa,

That's great. I would have been totally at a loss to explain it otherwise!

Thank you for running this test. I knew this new behavior was possible, but
I didn't expect it to happen on real data. The algorithm is essentially a
greedy algorithm, since at each step we assign the best possible available
reviewer to a paper (subject to a number of constraints). Because of this,
when the problem is heavily constrained, we might get to the end of the
algorithm and have no feasible assignments left (for example, if paper A
has been assigned reviewers 1 and 2 already, and we need to assign it a 3rd
reviewer, but the only reviewers still available are reviewers 1 and 2).
This is why the algorithm can fail.

With regards to the message you mentioned: We first try to run a version
where at each step we specifically ensure the WEF1 constraint (a fairness
constraint that must hold between every pair of papers). If that fails, we
run a less-constrained version. I thought that at least the second version
would always output an allocation, but it seems this isn't the case.

I think that the best thing to do in this case would be to either output a
failure message saying that the problem is too heavily constrained to use
FairSequence, or implicitly fall back to FairFlow or MinMax. I have my
meeting with Yair tomorrow, so I'll see if he has any better ideas.

Thanks,
Justin

On Wed, Apr 13, 2022 at 3:07 PM Melisa Bok ***@***.***> wrote:

> Hi @justinpayan <https://github.com/justinpayan> ,
>
> Apologies for this. This is my mistake and I didn't notice there were
> users with a reduce quota lower than the global min_papers value.
>
> However with @purujitgoyal <https://github.com/purujitgoyal> we were
> testing ICLR AC assignment and we couldn't find a solution using
> FairSequence while we found a solution using FairFlow and MinMax solvers.
> We don't know how to explain that.
>
> Something that we found in the logs is that the following message:
>
> Unable to find a WEF1 allocation satisfying all papers' demands. Falling back to picking sequence without WEF1 guarantees.
>
> Purujit can reproduce in his local environment so you may want to have a
> call and debug this together.
>
> Thanks,
>
> Melisa
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/openreview/openreview-matcher/pull/226#issuecomment-1098388703>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDA66IAGEC5PZMBJL53Y3DVE4LQ3ANCNFSM5ER5C7XQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 

Justin Payan

PhD Candidate

College of Information and Computer Sciences

University of Massachusetts Amherst

justinpayan commented:
Update on the failed run on ICLR: 

I believe I have implemented a solution to this problem. I will write a few more test cases and clean up the code, then I'll reach out to @purujitgoyal to check that this new approach works on ICLR. Thank you @melisabok for pointing out this interesting failure mode.

I am curious how good the solution will be, both in terms of overall score and fairness. But with this fix, the algorithm should at least return a valid allocation.

Thanks!
Justin

justinpayan commented:
I've implemented our solution to the problem discussed above on ICLR, and I added 4 unit tests to make sure that solution is working as intended. I'm looking forward to seeing if it solves the problem on ICLR. @purujitgoyal @melisabok

purujitgoyal commented:
@justinpayan I tested the new implementation and I was able to get the solution. However, there were very few users who were assigned less number of papers than the global `min_papers`. There was no `reduced_quota` set for these users this time. I just wanted to confirm is this intended as part of the solution?

justinpayan commented:
Hi @purujitgoyal, thank you very much for running that test. I think I know what the problem might be, and I'll see if I can fix it. I don't think it will be too difficult to fix. Thanks!

melisabok commented:
@purujitgoyal is this PR ready to merge? I would like the NeurIPS PCs to give it a try.
